internal/kubeclient/event.go

```diff
@@ -0,0 +1,139 @@
+package kubeclient
+
+import (
+	"context"
+	"errors"
+
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	libClient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/akuity/kargo/internal/logging"
+)
+
+type eventRecorder struct {
+	backoff wait.Backoff
+	sink    record.EventSink
+	logger  *log.Entry
+
+	newEventHandlerFn func(event *corev1.Event) func() error
+}
+
+// NewEventRecorder returns a new event recorder that records all events
+// without aggregation, even the given event is correlated.
+//
+// NOTE: This recorder must be used with caution as it creates a new Event
+// on every event without throttling / spam filtering features - which are
+// included in the event recorder from k8s.io/client-go. This could lead
+// to performance issues with the Kubernetes API server.
```

Comment:
I don't know a _ton_ about k8s events. This note gives me pause. Are we absolutely sure we're using it with the requisite caution?

Reply:
By default, Kubernetes API server will delete newly created events after 1 hours (configured by [`--event-ttl` flag](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/)).
I think it will be fine in most cases since we are trying to emit event once - so it depends on user's usage.

But we can talk about this more in separate issue for sure 🙏